docs: interpreting longtasks as spans (#597)
elastic/apm-agent-rum-js: docs/advanced-topics.asciidoc

[[advanced-topics]]
== Advanced Topics

* <<longtasks>>
* <<typescript>>
* <<custom-transactions, Custom Transactions>>
* <<custom-transaction-name, Initial page load transaction names>>

[[longtasks]]
=== How to interpret long task spans in the UI

Long tasks is a new performance metric that can be used for measuring the
responsiveness of an application and helps developers to understand the bad user
experience. It enables detecting tasks that monopolize the UI thread for
extended periods (greater than 50 milliseconds) and block other

> critical tasks from being executed as stated in the https://github.com/w3c/longtasks[official spec].

Reviewer: What if we swapped the what and the why? Something like:
```
Long tasks enables the detection of tasks that monopolize the
UI thread for extended periods of time (greater than 50 milliseconds) and block other
critical tasks from being executed -- as stated in the https://github.com/w3c/longtasks[official spec].
This new performance metric is useful for measuring the responsiveness of an application,
and can help developers better understand bad user experiences.
```

Author: Don't really oppose your suggestion here. But shouldn't we start with explaining what is long task and then go ahead with why its intereting?.

RUM agent automatically captures these Long tasks and include them as spans as
part of the transaction. Since long tasks currently does not have the full information on
which part of code cause slowness, it would be hard to interpret these spans.
Below you can find some tips to help with interpreting long task spans:

> + Name of the long task span (eg: `self`, `same-origin`) would imply the

Reviewer: ```suggestion
* The name of the long task span, e.g.: `self`, `same-origin`, etc., implies the
```

  origin of the task. It could be the current browsing context or inside iframes.

> + Context of the span is enriched with useful information like `attribution`

Reviewer: ```suggestion
* Context of the span is enriched with useful information like `attribution`
```

>   (the type of work such as script, layout, etc), `type`, `id` and `name` which

Reviewer: ```suggestion
  (the type of work, such as script, layout, etc), `type`, `id` and `name`, which
```

>   determines the culprit container(such as window, iframe, embed or object)

Reviewer: ```suggestion
  determines the culprit container (such as window, iframe, embed or object)
```

  responsible for the long task.

With the help of the transaction timeline and the span timings, one could dig
deeper by marking the slow application code using https://developer.mozilla.org/en-US/docs/Web/API/Performance/mark[User Timing
API] which would be again captured as spans by the agent and combining them along

> with long tasks would reveal the true source code location.

Reviewer: This paragraph is a little tough for me. What about something like this:
```
With the help of the transaction timeline and span timings,
one could dig deeper by marking slow application code with the
https://developer.mozilla.org/en-US/docs/Web/API/Performance/mark[User Timing API].
When these spans are then captured by the agent again, you could combine them
with long tasks to reveal the true source code location.
```

* <<custom-transactions, Custom Transactions>>
* <<custom-transaction-name, Initial page load transaction names>>
* <<longtasks>>

> * <<typescript>>

Reviewer: Kind of unrelated to this PR, but since we're touching this file, let's move the "custom-*" links to the bottom of the list and remove the custom titles so we match the order in the docs:
```
* <<longtasks>>
* <<typescript>>
* <<custom-transactions>>
* <<custom-transaction-name>>
```

[[longtasks]]
=== How to interpret Longtask Spans in the UI

> Longtasks is a new performance metric that helps measure the responsiveness and

Reviewer 2: ```suggestion
Long tasks is a new performance metric that can be used for measuring the responsiveness of an application and 
```

> helps understand the bad user experience on the web. It enables detecting tasks

Reviewer 2: ```suggestion
helps application developers to understand the user experience on their application. It enables detecting tasks
```

Author: changed this a bit since application was called in multiple places.

that monopolize the UI thread for extended periods (greater than 50 milliseconds) and block other

> critical tasks from being executed as stated in the official spec[https://github.com/w3c/longtasks].

Reviewer: ```suggestion
critical tasks from being executed as stated in the https://github.com/w3c/longtasks[official spec].
```

RUM agent automatically captures these Longtasks and include them as Spans as
part of the transaction. Since Longtasks currently does not have the full information on
which part of code cause slowness, it would be hard to interpret these spans.

> Some tips below would interpret these spans in a meaningful way

Reviewer 2: ```suggestion
Below you can find some tips to help with interpreting Longtask spans:
```

> + Name of the Longtask span could be `self`, `same-origin`, etc. would imply the

Reviewer 2: ```suggestion
+ Name of the Longtask span (e.g. `self`, `same-origin`) would imply the
```

  origin of the task. It could be the current browsing context or inside iframes.

+ Context of the span is enriched with useful information like `attribution`
  (the type of work such as script, layout, etc), `type`, `id` and `name` which
  determines the culprit container(such as window, iframe, embed or object)
  responsible for the long task

With the help of the transaction timeline and the span timings, one could dig
deeper by marking the slow application code using User Timing

> API[https://developer.mozilla.org/en-US/docs/Web/API/Performance/mark]

Reviewer: This, should be:
```
deeper by marking the slow application code using
https://developer.mozilla.org/en-US/docs/Web/API/Performance/mark[User Timing API],
```
